## 데이터 읽어오기

In [1]:
import pandas as pd
from pandas import DataFrame, Series

data_korea = pd.read_csv('총데이터/일별_대기질.csv') # 한국 대기질 데이터
data_korea.rename(columns = {'date':'Date'}, inplace = True)

data_korea

Date  SO2 (ppm)  PM10 (ug/m3)  O3 (ppm)  NO2 (ppm)  CO (ppm)  \
0     2017-06-01      0.001          42.0     0.024      0.023       0.5   
1     2017-06-02      0.002          34.0     0.030      0.023       0.6   
2     2017-06-03      0.003          25.0     0.032      0.021       0.5   
3     2017-06-04      0.003          31.0     0.040      0.020       0.5   
4     2017-06-05      0.004          35.0     0.033      0.031       0.7   
...          ...        ...           ...       ...        ...       ...   
1486  2021-06-26      0.003          19.0     0.036      0.018       0.4   
1487  2021-06-27      0.003          22.0     0.040      0.014       0.4   
1488  2021-06-28      0.003          24.0     0.038      0.018       0.4   
1489  2021-06-29      0.003          25.0     0.037      0.019       0.4   
1490  2021-06-30      0.003          38.0     0.034      0.024       0.4   

      PM2.5 (ug/m3)  
0              24.0  
1              14.0  
2              11.0  
3              14.0  
4              14.0  
...             ...  
1486           10.0  
1487           13.0  
1488           15.0  
1489           16.0  
1490           23.0  

[1491 rows x 7 columns]

In [2]:
data_B = pd.read_excel('총데이터/일별_베이징대기질.xlsx') # 베이징 대기질 데이터

data_B

date  PM2.5  PM10
0    2017-06-01     23    57
1    2017-06-02     23    65
2    2017-06-03     29    56
3    2017-06-04     48    93
4    2017-06-05     65   115
...         ...    ...   ...
1486 2021-06-26     22    40
1487 2021-06-27     26    48
1488 2021-06-28     32    53
1489 2021-06-29     26    40
1490 2021-06-30     14    27

[1491 rows x 3 columns]

In [3]:
data_J = pd.read_excel('총데이터/일별_지난대기질.xlsx') # 지난 대기질 데이터

data_J

date  PM2.5  PM10
0    2017-06-01     72   164
1    2017-06-02     44   132
2    2017-06-03     28    79
3    2017-06-04     48   114
4    2017-06-05     23    63
...         ...    ...   ...
1486 2021-06-26     26    44
1487 2021-06-27     38    66
1488 2021-06-28     43    75
1489 2021-06-29     22    40
1490 2021-06-30     24    46

[1491 rows x 3 columns]

In [4]:
data_T = pd.read_excel('총데이터/일별_텐진대기질.xlsx') # 텐진 대기질 데이터

data_T

date  PM2.5  PM10
0    2017-06-01     40    67
1    2017-06-02     14    52
2    2017-06-03     30    54
3    2017-06-04     35    70
4    2017-06-05     52    97
...         ...    ...   ...
1486 2021-06-26     32    61
1487 2021-06-27     37    70
1488 2021-06-28     43    77
1489 2021-06-29     27    46
1490 2021-06-30     27    44

[1491 rows x 3 columns]

In [5]:
data_weather = pd.read_csv('기상데이터/일별_문화동기상.csv', encoding = 'cp949') # 기상 데이터

data_weather

일시  평균기온(°C)  일강수량(mm)  평균 풍속(m/s)
0     2017-06-01      22.0       0.0         1.8
1     2017-06-02      20.6       0.0         1.5
2     2017-06-03      20.6       0.0         1.8
3     2017-06-04      21.2       0.0         1.4
4     2017-06-05      22.6       0.0         1.3
...          ...       ...       ...         ...
1486  2021-06-26      22.7      11.5         0.9
1487  2021-06-27      24.9       0.0         0.9
1488  2021-06-28      25.1       0.0         1.2
1489  2021-06-29      23.7       0.0         1.0
1490  2021-06-30      24.6       9.0         0.9

[1491 rows x 4 columns]

In [6]:
data_weather[data_weather['일시'].duplicated()] # 중복되는 값 확인

Empty DataFrame
Columns: [일시, 평균기온(°C), 일강수량(mm), 평균 풍속(m/s)]
Index: []

In [7]:
data_traffic = pd.read_excel('총데이터/일별_대전교통량(수정).xlsx') # 대전 교통량 데이터
# data_traffic.drop(columns=['Unnamed: 0'], inplace=True)
data_traffic.rename(columns = {'평균':'교통량합계'}, inplace = True)

data_traffic

연도  월   일     교통량합계
0     2017  6   1  108656.0
1     2017  6   2  114477.0
2     2017  6   3  111557.0
3     2017  6   4   57121.0
4     2017  6   5  112199.0
...    ... ..  ..       ...
1486  2021  6  26   91068.0
1487  2021  6  27   76373.0
1488  2021  6  28   97692.0
1489  2021  6  29   96636.0
1490  2021  6  30   97579.0

[1491 rows x 4 columns]

## 데이터 합치기

In [8]:
# 구분 쉽게 columns 이름 바꿔놓기

data_T.rename(columns = {'PM2.5':'TPM2.5'}, inplace = True)
data_B.rename(columns = {'PM2.5':'BPM2.5'}, inplace = True)
data_J.rename(columns = {'PM2.5':'JPM2.5'}, inplace = True)
data_T.rename(columns = {'PM10':'TPM10'}, inplace = True)
data_B.rename(columns = {'PM10':'BPM10'}, inplace = True)
data_J.rename(columns = {'PM10':'JPM10'}, inplace = True)

In [9]:
data = data_T.merge(data_J)
data = data.merge(data_B)

data

date  TPM2.5  TPM10  JPM2.5  JPM10  BPM2.5  BPM10
0    2017-06-01      40     67      72    164      23     57
1    2017-06-02      14     52      44    132      23     65
2    2017-06-03      30     54      28     79      29     56
3    2017-06-04      35     70      48    114      48     93
4    2017-06-05      52     97      23     63      65    115
...         ...     ...    ...     ...    ...     ...    ...
1486 2021-06-26      32     61      26     44      22     40
1487 2021-06-27      37     70      38     66      26     48
1488 2021-06-28      43     77      43     75      32     53
1489 2021-06-29      27     46      22     40      26     40
1490 2021-06-30      27     44      24     46      14     27

[1491 rows x 7 columns]

In [10]:
data = pd.concat([data, data_traffic], axis=1)
data = pd.concat([data, data_weather], axis=1)
data = pd.concat([data, data_korea], axis=1)

In [11]:
data.drop(columns=['일시', '월', '일', '연도', 'Date'], inplace=True)

data

date  TPM2.5  TPM10  JPM2.5  JPM10  BPM2.5  BPM10     교통량합계  \
0    2017-06-01      40     67      72    164      23     57  108656.0   
1    2017-06-02      14     52      44    132      23     65  114477.0   
2    2017-06-03      30     54      28     79      29     56  111557.0   
3    2017-06-04      35     70      48    114      48     93   57121.0   
4    2017-06-05      52     97      23     63      65    115  112199.0   
...         ...     ...    ...     ...    ...     ...    ...       ...   
1486 2021-06-26      32     61      26     44      22     40   91068.0   
1487 2021-06-27      37     70      38     66      26     48   76373.0   
1488 2021-06-28      43     77      43     75      32     53   97692.0   
1489 2021-06-29      27     46      22     40      26     40   96636.0   
1490 2021-06-30      27     44      24     46      14     27   97579.0   

      평균기온(°C)  일강수량(mm)  평균 풍속(m/s)  SO2 (ppm)  PM10 (ug/m3)  O3 (ppm)  \
0         22.0       0.0         1.8      0.001          42.0     0.024   
1         20.6       0.0         1.5      0.002          34.0     0.030   
2         20.6       0.0         1.8      0.003          25.0     0.032   
3         21.2       0.0         1.4      0.003          31.0     0.040   
4         22.6       0.0         1.3      0.004          35.0     0.033   
...        ...       ...         ...        ...           ...       ...   
1486      22.7      11.5         0.9      0.003          19.0     0.036   
1487      24.9       0.0         0.9      0.003          22.0     0.040   
1488      25.1       0.0         1.2      0.003          24.0     0.038   
1489      23.7       0.0         1.0      0.003          25.0     0.037   
1490      24.6       9.0         0.9      0.003          38.0     0.034   

      NO2 (ppm)  CO (ppm)  PM2.5 (ug/m3)  
0         0.023       0.5           24.0  
1         0.023       0.6           14.0  
2         0.021       0.5           11.0  
3         0.020       0.5           14.0  
4         0.031       0.7           14.0  
...         ...       ...            ...  
1486      0.018       0.4           10.0  
1487      0.014       0.4           13.0  
1488      0.018       0.4           15.0  
1489      0.019       0.4           16.0  
1490      0.024       0.4           23.0  

[1491 rows x 17 columns]

In [12]:
data.to_csv('총데이터/일별_전체데이터.csv', encoding = 'cp949', index=False)

In [322]:
data_nodate = data.drop(columns=['date'])

data_nodate

TPM2.5  TPM10  JPM2.5  JPM10  BPM2.5  BPM10     교통량합계  평균기온(°C)  \
0         40     67      72    164      23     57  108656.0      22.0   
1         14     52      44    132      23     65  114477.0      20.6   
2         30     54      28     79      29     56  111557.0      20.6   
3         35     70      48    114      48     93   57121.0      21.2   
4         52     97      23     63      65    115  112199.0      22.6   
...      ...    ...     ...    ...     ...    ...       ...       ...   
1486      32     61      26     44      22     40   91068.0      22.7   
1487      37     70      38     66      26     48   76373.0      24.9   
1488      43     77      43     75      32     53   97692.0      25.1   
1489      27     46      22     40      26     40   96636.0      23.7   
1490      27     44      24     46      14     27   97579.0      24.6   

      일강수량(mm)  평균 풍속(m/s)  SO2 (ppm)  PM10 (ug/m3)  O3 (ppm)  NO2 (ppm)  \
0          0.0         1.8      0.001          42.0     0.024      0.023   
1          0.0         1.5      0.002          34.0     0.030      0.023   
2          0.0         1.8      0.003          25.0     0.032      0.021   
3          0.0         1.4      0.003          31.0     0.040      0.020   
4          0.0         1.3      0.004          35.0     0.033      0.031   
...        ...         ...        ...           ...       ...        ...   
1486      11.5         0.9      0.003          19.0     0.036      0.018   
1487       0.0         0.9      0.003          22.0     0.040      0.014   
1488       0.0         1.2      0.003          24.0     0.038      0.018   
1489       0.0         1.0      0.003          25.0     0.037      0.019   
1490       9.0         0.9      0.003          38.0     0.034      0.024   

      CO (ppm)  PM2.5 (ug/m3)  
0          0.5           24.0  
1          0.6           14.0  
2          0.5           11.0  
3          0.5           14.0  
4          0.7           14.0  
...        ...            ...  
1486       0.4           10.0  
1487       0.4           13.0  
1488       0.4           15.0  
1489       0.4           16.0  
1490       0.4           23.0  

[1491 rows x 16 columns]

In [323]:
data_nodate = data_nodate.set_axis(['TPM2.5', 'TPM10', 'JPM2.5', 'JPM10', 'BPM2.5', 'BPM10', 
                                   '교통량합계', '평균기온', '일강수량', '평균풍속', 
                                   'SO2', 'PM10', 'O3', 'NO2', 'CO', 'PM2.5'], axis=1)

data_nodate

TPM2.5  TPM10  JPM2.5  JPM10  BPM2.5  BPM10     교통량합계  평균기온  일강수량  평균풍속  \
0         40     67      72    164      23     57  108656.0  22.0   0.0   1.8   
1         14     52      44    132      23     65  114477.0  20.6   0.0   1.5   
2         30     54      28     79      29     56  111557.0  20.6   0.0   1.8   
3         35     70      48    114      48     93   57121.0  21.2   0.0   1.4   
4         52     97      23     63      65    115  112199.0  22.6   0.0   1.3   
...      ...    ...     ...    ...     ...    ...       ...   ...   ...   ...   
1486      32     61      26     44      22     40   91068.0  22.7  11.5   0.9   
1487      37     70      38     66      26     48   76373.0  24.9   0.0   0.9   
1488      43     77      43     75      32     53   97692.0  25.1   0.0   1.2   
1489      27     46      22     40      26     40   96636.0  23.7   0.0   1.0   
1490      27     44      24     46      14     27   97579.0  24.6   9.0   0.9   

        SO2  PM10     O3    NO2   CO  PM2.5  
0     0.001  42.0  0.024  0.023  0.5   24.0  
1     0.002  34.0  0.030  0.023  0.6   14.0  
2     0.003  25.0  0.032  0.021  0.5   11.0  
3     0.003  31.0  0.040  0.020  0.5   14.0  
4     0.004  35.0  0.033  0.031  0.7   14.0  
...     ...   ...    ...    ...  ...    ...  
1486  0.003  19.0  0.036  0.018  0.4   10.0  
1487  0.003  22.0  0.040  0.014  0.4   13.0  
1488  0.003  24.0  0.038  0.018  0.4   15.0  
1489  0.003  25.0  0.037  0.019  0.4   16.0  
1490  0.003  38.0  0.034  0.024  0.4   23.0  

[1491 rows x 16 columns]

In [325]:
new_index = ['PM2.5', 'PM10', '교통량합계', '평균기온', '일강수량', '평균풍속',
             'TPM2.5', 'TPM10', 'BPM2.5', 'BPM10', 'JPM2.5', 'JPM10', 'SO2', 'O3', 'NO2', 'CO']

data_nodate = data_nodate.reindex(new_index, axis = 1)

data_nodate

PM2.5  PM10     교통량합계  평균기온  일강수량  평균풍속  TPM2.5  TPM10  BPM2.5  BPM10  \
0      24.0  42.0  108656.0  22.0   0.0   1.8      40     67      23     57   
1      14.0  34.0  114477.0  20.6   0.0   1.5      14     52      23     65   
2      11.0  25.0  111557.0  20.6   0.0   1.8      30     54      29     56   
3      14.0  31.0   57121.0  21.2   0.0   1.4      35     70      48     93   
4      14.0  35.0  112199.0  22.6   0.0   1.3      52     97      65    115   
...     ...   ...       ...   ...   ...   ...     ...    ...     ...    ...   
1486   10.0  19.0   91068.0  22.7  11.5   0.9      32     61      22     40   
1487   13.0  22.0   76373.0  24.9   0.0   0.9      37     70      26     48   
1488   15.0  24.0   97692.0  25.1   0.0   1.2      43     77      32     53   
1489   16.0  25.0   96636.0  23.7   0.0   1.0      27     46      26     40   
1490   23.0  38.0   97579.0  24.6   9.0   0.9      27     44      14     27   

      JPM2.5  JPM10    SO2     O3    NO2   CO  
0         72    164  0.001  0.024  0.023  0.5  
1         44    132  0.002  0.030  0.023  0.6  
2         28     79  0.003  0.032  0.021  0.5  
3         48    114  0.003  0.040  0.020  0.5  
4         23     63  0.004  0.033  0.031  0.7  
...      ...    ...    ...    ...    ...  ...  
1486      26     44  0.003  0.036  0.018  0.4  
1487      38     66  0.003  0.040  0.014  0.4  
1488      43     75  0.003  0.038  0.018  0.4  
1489      22     40  0.003  0.037  0.019  0.4  
1490      24     46  0.003  0.034  0.024  0.4  

[1491 rows x 16 columns]

In [326]:
# 대전 대기질 데이터에서 미세먼지만 가져오기
data_prac = data_nodate.drop(columns=['SO2', 'O3', 'NO2', 'CO'])

data_prac #'SO2', 'O3', 'NO2', 'CO'에 null값이 있기에 잠시 제외하기

PM2.5  PM10     교통량합계  평균기온  일강수량  평균풍속  TPM2.5  TPM10  BPM2.5  BPM10  \
0      24.0  42.0  108656.0  22.0   0.0   1.8      40     67      23     57   
1      14.0  34.0  114477.0  20.6   0.0   1.5      14     52      23     65   
2      11.0  25.0  111557.0  20.6   0.0   1.8      30     54      29     56   
3      14.0  31.0   57121.0  21.2   0.0   1.4      35     70      48     93   
4      14.0  35.0  112199.0  22.6   0.0   1.3      52     97      65    115   
...     ...   ...       ...   ...   ...   ...     ...    ...     ...    ...   
1486   10.0  19.0   91068.0  22.7  11.5   0.9      32     61      22     40   
1487   13.0  22.0   76373.0  24.9   0.0   0.9      37     70      26     48   
1488   15.0  24.0   97692.0  25.1   0.0   1.2      43     77      32     53   
1489   16.0  25.0   96636.0  23.7   0.0   1.0      27     46      26     40   
1490   23.0  38.0   97579.0  24.6   9.0   0.9      27     44      14     27   

      JPM2.5  JPM10  
0         72    164  
1         44    132  
2         28     79  
3         48    114  
4         23     63  
...      ...    ...  
1486      26     44  
1487      38     66  
1488      43     75  
1489      22     40  
1490      24     46  

[1491 rows x 12 columns]

In [502]:
data_prac.to_csv('총데이터/일별_모델데이터.csv', encoding = 'cp949', index=False)

## MLP 실습

### 표준화되지 않은 전체 데이터

In [494]:
X = data_prac.drop(columns=['PM2.5', 'PM10'], axis=1)

X

교통량합계  평균기온  일강수량  평균풍속  TPM2.5  TPM10  BPM2.5  BPM10  JPM2.5  JPM10
0     108656.0  22.0   0.0   1.8      40     67      23     57      72    164
1     114477.0  20.6   0.0   1.5      14     52      23     65      44    132
2     111557.0  20.6   0.0   1.8      30     54      29     56      28     79
3      57121.0  21.2   0.0   1.4      35     70      48     93      48    114
4     112199.0  22.6   0.0   1.3      52     97      65    115      23     63
...        ...   ...   ...   ...     ...    ...     ...    ...     ...    ...
1486   91068.0  22.7  11.5   0.9      32     61      22     40      26     44
1487   76373.0  24.9   0.0   0.9      37     70      26     48      38     66
1488   97692.0  25.1   0.0   1.2      43     77      32     53      43     75
1489   96636.0  23.7   0.0   1.0      27     46      26     40      22     40
1490   97579.0  24.6   9.0   0.9      27     44      14     27      24     46

[1491 rows x 10 columns]

In [495]:
#null_counts = X.isnull().sum()
#null_counts[null_counts > 0] # NULL값이 없음을 확인

In [496]:
# 학습 데이터 지정하기

import numpy as np
import sklearn.model_selection as ms

# X = np.array(data_korea['PM2.5 (ug/m3)']).reshape(-1, 1)
# X = X.to_numpy()
y = data_prac['PM2.5'] # 'PM10'

X_train, X_test, y_train, y_test = ms.train_test_split(X, y, test_size = 0.2, random_state = 1)
# X.shape

In [497]:
'''
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()   # 객체 만들기
scaler.fit(X_train)     # 변환 규칙을 익히기
X_train_scaled = scaler.transform(X_train)  # 데이터를 표준화 전처리
X_test_scaled = scaler.transform(X_test)      # 데이터를 표준화 전처리
'''

'\nfrom sklearn.preprocessing import StandardScaler\n\nscaler = StandardScaler()   # 객체 만들기\nscaler.fit(X_train)     # 변환 규칙을 익히기\nX_train_scaled = scaler.transform(X_train)  # 데이터를 표준화 전처리\nX_test_scaled = scaler.transform(X_test)      # 데이터를 표준화 전처리\n'

In [498]:
clf = nn.MLPClassifier(hidden_layer_sizes=(10,), activation='tanh', \
                    solver='adam', alpha=0.01, batch_size=32, \
                    learning_rate_init=0.1, max_iter=500)  # 객체 생성

In [500]:
# 학습하기

mlp = clf.fit(X_train, y_train)

y_pred = mlp.predict(X_test)

In [ ]:
clf1 = nn.MLPClassifier(hidden_layer_sizes=(10,), activation='sigmoid', \
                       batch_size=32,  max_iter=500)  # 객체 생성

In [ ]:
# 학습하기2

mlp = clf.fit(X_train, y_train)

y_pred = mlp.predict(X_test)

In [501]:
print("train:", mlp.score(X_train, y_train))# scaled 말고 기본
print("test:", mlp.score(X_test, y_test))

train: 0.03523489932885906
test: 0.033444816053511704


In [470]:
from sklearn.metrics import mean_squared_error, r2_score

y_pred = clf.predict(X_test)

#MSE
mse = mean_squared_error(y_test, y_pred)

#RMSE
rmse = np.sqrt(mse)

print("MSE: ", np.round(mse, 3))
print("RMSE: ", np.round(rmse, 3))
print("정확도=", mt.accuracy_score(y_test, y_pred))

MSE:  259.763
RMSE:  16.117
정확도= 0.049107142857142856


In [471]:
#결정계수 R2
r2 = r2_score(y_test, y_pred)
print("R2: ", np.round(r2, 3))

R2:  -0.296


logistic : 0.0379

relu : 0.049

tanh : 0.0446

with hidden_layer_size(100,) : 0.0379

without learning_rate : 0.0491

without learning_rate init : 0.040

In [ ]:
# mlp = nn.MLPClassifier(max_iter=500) GridSarchCV 활용용

## GridSearchCV 활용

mlp.get_params().keys() :

dict_keys(['activation', 'alpha', 'batch_size', 'beta_1', 'beta_2', 'early_stopping', 'epsilon', 'hidden_layer_sizes', 'learning_rate', 'learning_rate_init', 'max_fun', 'max_iter', 'momentum', 'n_iter_no_change', 'nesterovs_momentum', 'power_t', 'random_state', 'shuffle', 'solver', 'tol', 'validation_fraction', 'verbose', 'warm_start'])

In [439]:
from sklearn.model_selection import GridSearchCV

parameter_space = {
    'hidden_layer_sizes': [(10,),(20,)], #[(50,50,50), (50,100,50), (100,)], or[(10,30,10),(20,)]
    'activation': ['tanh', 'relu', 'logistic'],
    'solver': ['sgd', 'adam'],
    'alpha': [0.0001, 0.05],
    'learning_rate': ['constant','adaptive'],
}

In [440]:
clf = GridSearchCV(mlp, param_grid = parameter_space, n_jobs=-1, cv = 5)
clf.fit(X_train, y_train)

C:\Users\567di\anaconda3\lib\site-packages\sklearn\model_selection\_split.py:666: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(("The least populated class in y has only %d"


GridSearchCV(cv=5, estimator=MLPClassifier(max_iter=500), n_jobs=-1,
             param_grid={'activation': ['tanh', 'relu', 'logistic'],
                         'alpha': [0.0001, 0.05],
                         'hidden_layer_sizes': [(10,), (20,)],
                         'learning_rate': ['constant', 'adaptive'],
                         'solver': ['sgd', 'adam']})

In [441]:
print(clf.score(X_train, y_train))
print(clf.score(X_test, y_test))

0.045062320230105465
0.049107142857142856


In [442]:
# Best parameters
print('Best parameters found:\n', clf.best_params_)

Best parameters found:
 {'activation': 'tanh', 'alpha': 0.0001, 'hidden_layer_sizes': (20,), 'learning_rate': 'constant', 'solver': 'adam'}


In [443]:
# All results
means = clf.cv_results_['mean_test_score']
stds = clf.cv_results_['std_test_score']
for mean, std, params in zip(means, stds, clf.cv_results_['params']):
    print("%0.3f (+/-%0.03f) for %r" % (mean, std * 2, params))

0.044 (+/-0.004) for {'activation': 'tanh', 'alpha': 0.0001, 'hidden_layer_sizes': (10,), 'learning_rate': 'constant', 'solver': 'sgd'}
0.043 (+/-0.000) for {'activation': 'tanh', 'alpha': 0.0001, 'hidden_layer_sizes': (10,), 'learning_rate': 'constant', 'solver': 'adam'}
0.043 (+/-0.000) for {'activation': 'tanh', 'alpha': 0.0001, 'hidden_layer_sizes': (10,), 'learning_rate': 'adaptive', 'solver': 'sgd'}
0.042 (+/-0.004) for {'activation': 'tanh', 'alpha': 0.0001, 'hidden_layer_sizes': (10,), 'learning_rate': 'adaptive', 'solver': 'adam'}
0.043 (+/-0.000) for {'activation': 'tanh', 'alpha': 0.0001, 'hidden_layer_sizes': (20,), 'learning_rate': 'constant', 'solver': 'sgd'}
0.045 (+/-0.005) for {'activation': 'tanh', 'alpha': 0.0001, 'hidden_layer_sizes': (20,), 'learning_rate': 'constant', 'solver': 'adam'}
0.043 (+/-0.000) for {'activation': 'tanh', 'alpha': 0.0001, 'hidden_layer_sizes': (20,), 'learning_rate': 'adaptive', 'solver': 'sgd'}
0.043 (+/-0.000) for {'activation': 'tanh', '

In [444]:
print('Best score is:\n', clf.best_score_)

Best score is:
 0.04505888847994112


## MLP 여러 종류의 파라미터로 시도

### without weather parameters (표준화된 데이터로 시도)

In [424]:
X1 = data_prac.drop(columns=['PM2.5', 'PM10', '평균기온', '일강수량', '평균풍속'], axis=1)

y = data_prac['PM2.5'] # 'PM10'

X_train1, X_test1, y_train1, y_test1 = ms.train_test_split(X1, y, test_size = 0.3, random_state=100)

In [425]:
scaler = StandardScaler()   # 객체 만들기
scaler.fit(X_train1)     # 변환 규칙을 익히기
X_train_scaled1 = scaler.transform(X_train1)  
X_test_scaled1 = scaler.transform(X_test1)      # 데이터를 표준화 전처리

In [426]:
mlp1 = nn.MLPClassifier(hidden_layer_sizes=(10,), activation='logistic', \
                    solver='sgd', alpha=0.01, batch_size=32, \
                    learning_rate_init=0.1, max_iter=500)  # 객체 생성

In [427]:
parameter_space = {
    'hidden_layer_sizes': [(10,),(20,)], #[(50,50,50), (50,100,50), (100,)], #[(10,30,10),(20,)]
    'activation': ['tanh', 'relu'],
    'solver': ['sgd', 'adam'],
    'alpha': [0.0001, 0.05],
    'learning_rate': ['constant','adaptive'],
}

In [428]:
clf1 = GridSearchCV(mlp1, param_grid = parameter_space, n_jobs=-1, cv = 5) #cv = 3로도 시도
clf1.fit(X_train_scaled1, y_train1) # scaled 말고 기본으로 시도

C:\Users\567di\anaconda3\lib\site-packages\sklearn\model_selection\_split.py:666: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(("The least populated class in y has only %d"


GridSearchCV(cv=5,
             estimator=MLPClassifier(activation='logistic', alpha=0.01,
                                     batch_size=32, hidden_layer_sizes=(10,),
                                     learning_rate_init=0.1, max_iter=500,
                                     solver='sgd'),
             n_jobs=-1,
             param_grid={'activation': ['tanh', 'relu'],
                         'alpha': [0.0001, 0.05],
                         'hidden_layer_sizes': [(10,), (20,)],
                         'learning_rate': ['constant', 'adaptive'],
                         'solver': ['sgd', 'adam']})

In [429]:
print("train:", clf1.score(X_train_scaled1, y_train1))# scaled 말고 기본
print("test:", clf1.score(X_test_scaled1, y_test1))

train: 0.05560882070949185
test: 0.017857142857142856


In [430]:
# Best parameters
print('Best parameters found:\n', clf1.best_params_)

Best parameters found:
 {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': (20,), 'learning_rate': 'constant', 'solver': 'adam'}


In [431]:
# Best score
print('Best score is:\n', clf1.best_score_)

Best score is:
 0.049861980125138025


### PM2.5

#### with weather (unscaled)
    
- 'hidden_layer_sizes': [(50,50,50), (50,100,50), (100,)]
    - Best score is 0.04505888847994112
    - Best parameters found is {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': (100,), 'learning_rate': 'adaptive', 'solver': 'sgd'}
    - train:0.045062320230105465
    - test:0.03571428571428571

#### without weather (scaled)

- 'hidden_layer_sizes': [(10,),(20,)]
    - Best score is 0.06328211262421789
    - Best parameters found is {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes':(10,), 'learning_rate': 'constant', 'solver': 'sgd'}
    - train: 0.12464046021093
    - test: 0.03571428571428571

- 'hidden_layer_sizes': [(10,),(20,)] with cv = 3 : 
    - Best score is 0.046015656916978585
    - Best parameters found is {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': (10,), 'learning_rate': 'constant', 'solver': 'adam'}
    - train: 0.04793863854266539
    - test: 0.03571428571428571
    
- 'hidden_layer_sizes': [(10,30,10),(20,)]
    - Best score is 0.05080971659919028
    - Best parameters found is {'activation': 'relu', 'alpha': 0.05, 'hidden_layer_sizes': (20,),
      'learning_rate': 'adaptive', 'solver': 'sgd'}
    - train: 0.18216682646212848
    - test: 0.03125
    
- 'hidden_layer_sizes': [(50,50,50), (50,100,50), (100,)]
    - Best score is 0.04890044166359956
    - Best parameters found is {'activation': 'relu', 'alpha': 0.05, 'hidden_layer_sizes': (100,), 'learning_rate': 'adaptive', 'solver': 'sgd'}
    - train: 0.3451581975071908
    - test: 0.03794642857142857
    
#### without weather (unscaled)

- 'hidden_layer_sizes': [(10,),(20,)]: (random_state=1)
    - Best score is 0.04410195068089805
    - Best parameters found is {'activation': 'relu', 'alpha': 0.05, 'hidden_layer_sizes': (100,), 'learning_rate': 'adaptive', 'solver': 'sgd'}
    - train: 0.045062320230105465
    - test: 0.049107142857142856
    
- 'hidden_layer_sizes': [(10,),(20,)]: (random_state=100)
    - Best score is 0.049861980125138025
    - Best parameters found is {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': (20,), 'learning_rate': 'constant', 'solver': 'adam'}
    - train: 0.05560882070949185
    - test: 0.017857142857142856  

## 연습

In [269]:
import sklearn.neural_network as nn

ppn_clf = nn.MLPClassifier(learning_rate_init=0.1, max_iter=1000, activation='logistic',
                           hidden_layer_sizes=(10,)) # 50,40,30

In [270]:
# 학습하기

clf = ppn_clf.fit(X_train, y_train)

y_pred = clf.predict(X_test)

In [271]:
print(np.round(y_pred,2))

[11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11.
 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11.
 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11.
 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11.
 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11.
 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11.
 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11.
 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11.
 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11.
 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11.
 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11.
 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11.
 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11.
 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11. 11

In [272]:
import sklearn.metrics as mt

# score = mt.accuracy_score(y_test, y_pred) # 정답과 예측값을 순서대로 적기

print("정확도 =", score)
print(mt.confusion_matrix(y_test, y_pred))
print()
print(mt.classification_report(y_test, y_pred))

정확도 = 0.03571428571428571
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]

              precision    recall  f1-score   support

         3.0       0.00      0.00      0.00         1
         4.0       0.00      0.00      0.00         3
         5.0       0.00      0.00      0.00         9
         6.0       0.00      0.00      0.00         9
         7.0       0.00      0.00      0.00        18
         8.0       0.00      0.00      0.00         7
         9.0       0.00      0.00      0.00        14
        10.0       0.00      0.00      0.00        25
        11.0       0.04      1.00      0.09        20
        12.0       0.00      0.00      0.00        16
        13.0       0.00      0.00      0.00        19
        14.0       0.00      0.00      0.00        22
        15.0       0.00      0.00      0.00        19
        16.0       0.00      0.00      0.00        13
        17.0       0.00      0.00      0.0

C:\Users\567di\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\567di\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\567di\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
clf = GridSearchCV(mlp, param_grid = parameter_space, n_jobs=-1, cv = 5)
clf.fit(X_train, y_train)